boa_engine/src/builtins/promise/mod.rs

```diff
@@ -0,0 +1,731 @@
+        };
+
+        // 7. Let stepsReject be the algorithm steps defined in Promise Reject Functions.
+        // 8. Let lengthReject be the number of non-optional parameters of the function definition in Promise Reject Functions.
+        // 9. Let reject be CreateBuiltinFunction(stepsReject, lengthReject, "", « [[Promise]], [[AlreadyResolved]] »).
+        let reject = FunctionBuilder::closure_with_captures(
+            context,
+            |this, args, captures, context| {
+                // https://tc39.es/ecma262/#sec-promise-reject-functions
+
+                // 1. Let F be the active function object.
+                // 2. Assert: F has a [[Promise]] internal slot whose value is an Object.
+                // 3. Let promise be F.[[Promise]].
+                // 4. Let alreadyResolved be F.[[AlreadyResolved]].
+                let RejectResolveCaptures {
+                    promise,
+                    already_resolved,
+                } = captures;
+
+                // 5. If alreadyResolved.[[Value]] is true, return undefined.
+                if already_resolved
+                    .get("Value", context)?
+                    .as_boolean()
+                    .unwrap_or(false)
+                {
+                    return Ok(JsValue::Undefined);
+                }
+
+                // 6. Set alreadyResolved.[[Value]] to true.
+                already_resolved.set("Value", true, true, context)?;
+
+                let reason = args.get_or_undefined(0);
+                // 7. Perform RejectPromise(promise, reason).
+                promise
+                    .borrow_mut()
+                    .as_promise_mut()
+                    .expect("Expected promise to be a Promise")
+                    .reject(reason, context)?;
+
+                // 8. Return undefined.
+                Ok(JsValue::Undefined)
+            },
+            reject_captures,
+        )
+        .name("")
+        .length(1)
+        .constructor(false)
+        .build();
+
+        // 10. Set reject.[[Promise]] to promise.
+        reject.set("Promise", promise.clone(), true, context)?;
+
+        // 11. Set reject.[[AlreadyResolved]] to alreadyResolved.
+        reject.set("AlreadyResolved", already_resolved, true, context)?;
+
+        // 12. Return the Record { [[Resolve]]: resolve, [[Reject]]: reject }.
+        let resolve = resolve.conv::<JsValue>();
+        let reject = reject.conv::<JsValue>();
+        Ok(ResolvingFunctionsRecord { resolve, reject })
+    }
+
+    /// More information:
+    ///  - [ECMAScript reference][spec]
+    ///
+    /// [spec]: https://tc39.es/ecma262/#sec-fulfillpromise
+    pub fn fulfill(&mut self, value: &JsValue, context: &mut Context) -> JsResult<()> {
+        // 1. Assert: The value of promise.[[PromiseState]] is pending.
+        match self.promise_state {
+            PromiseState::Pending => (),
+            _ => return context.throw_error("Expected promise.[[PromiseState]] to be pending"),
+        }
```

Comment:
An assertion should be asserted with `assert!` or `debug_assert!`  which more clearly indicates that it's an engine error and not an user error.

Reply:
This has been fixed in #2107.